src/zjs_net.c

```diff
@@ -0,0 +1,1111 @@
+            handle->wptr += len;
+
+            // if not paused, call the callback to get JS the data
+            if (!handle->paused) {
+                handle->tcp_read_id = zjs_add_c_callback(handle, tcp_c_callback);
+                zjs_signal_callback(handle->tcp_read_id, NULL, 0);
+
+                DBG_PRINT("data recieved on context %p: data=%p, len=%u\n", context, data, len);
+            }
+        }
+
+        net_buf_unref(buf);
+    }
+}
+
+static inline void pkt_sent(struct net_context *context,
+                            int status,
+                            void *token,
+                            void *user_data)
+{
+    if (!status) {
+        int sent = POINTER_TO_UINT(token);
+        DBG_PRINT("Sent %d bytes\n", sent);
+        if (sent) {
+            zjs_callback_id id = POINTER_TO_INT(user_data);
+            if (id != -1) {
+                zjs_signal_callback(id, NULL, 0);
+            }
+        }
+    }
+}
+
+/**
+ * Write data to a socket
+ *
+ * @name write
+ * @memberof Net.Socket
+ * @param {Buffer} buf - Buffer being written to the socket
+ * @param {function=} func - Callback called when write has completed
+ */
+static jerry_value_t socket_write(const jerry_value_t function_obj,
+                                  const jerry_value_t this,
+                                  const jerry_value_t argv[],
+                                  const jerry_length_t argc)
+{
+    ZJS_VALIDATE_ARGS(Z_OBJECT, Z_OPTIONAL Z_FUNCTION);
+
+    sock_handle_t *handle = NULL;
+
+    jerry_get_object_native_handle(this, (uintptr_t *)&handle);
+    if (!handle) {
+        return zjs_error("socket handle not found");
+    }
+    start_socket_timeout(handle, handle->timeout);
+
+    zjs_buffer_t *buf = zjs_buffer_find(argv[0]);
+    struct net_buf *send_buf;
+    send_buf = net_nbuf_get_tx(handle->tcp_sock, K_NO_WAIT);
+    if (!send_buf) {
+        ERR_PRINT("cannot acquire send_buf\n");
+        return jerry_create_boolean(false);
+    }
+
+    bool status = net_nbuf_append(send_buf, buf->bufsize, buf->buffer, K_NO_WAIT);
+    if (!status) {
+        net_nbuf_unref(send_buf);
+        ERR_PRINT("cannot populate send_buf\n");
+        return jerry_create_boolean(false);
+    }
+
+    zjs_callback_id id = -1;
+    if (argc > 1) {
+        id = zjs_add_callback_once(argv[1], this, NULL, NULL);
+    }
+    int ret = net_context_send(send_buf, pkt_sent,
+                               K_NO_WAIT,
+                               UINT_TO_POINTER(net_buf_frags_len(send_buf)),
+                               INT_TO_POINTER((int32_t)id));
+    if (ret < 0) {
+        ERR_PRINT("Cannot send data to peer (%d)", ret);
+        net_nbuf_unref(send_buf);
+        zjs_remove_callback(id);
+        // TODO: may need to check the specific error to determine action
+        DBG_PRINT("closing socket, context=%p, socket=%u\n", handle->tcp_sock, handle->socket);
+        zjs_trigger_event(handle->socket, "close", NULL, 0, post_closed, handle);
```

Comment:
Yeah I don't think this makes sense. A write failed, you're returning false, don't pretend it closed?